File: MyMonitorHub.Agent/MyMonitorHub.Agent.Core/Core/EventReportingThread.cs
using System;
using System.Diagnostics;
using System.Threading;
using System.Xml;
using MyMonitorHub.Common.WebApi;
using MyMonitorHub.Agent.Common;
using Serilog;

namespace MyMonitorHub.Agent.Core
{
    /// <summary>
    ///     Checks on the queue of events every minute and sends what it finds out to the monitoring
    ///     server.
    /// </summary>
    internal sealed class EventReportingThread : IThreadShutdown
    {
        private static readonly ILogger Logger = Log.ForContext<EventReportingThread>();


        private volatile bool _stopped = true;
        private readonly object _padlock = new object();

        private DateTime _lastSuccess;


        private void Run()
        {
            _lastSuccess = DateTime.Now;
            Logger.Information("Start of Event reporting thread");

            _stopped = false;
            while (!_stopped)
            {
                // wait 60 seconds before starting
                try
                {
                    lock (_padlock)
                    {
                        System.Threading.Monitor.Wait(_padlock, 60000);
                    }
                    if (_stopped)
                    {
                        break;
                    }
                }
                catch (Exception)
                {
                    _stopped = true;
                    break;
                }

                var start = DateTime.Now;

                var eventList = Monitor.EventList; // synced and creates a new empy list
                if (eventList.Count != 0)
                {
                    var packet = new EventPacket
                        {
                            Events = eventList.ToArray()
                        };

                    try
                    {
                        var service = new WebApiCall(MonitorProfile.Current.BaseAddress, MonitorProfile.TokenManager);
                        service.Send<bool>(MonitorProfile.Current.AccountId,
                            MonitorProfile.Current.DeviceId,
                            packet);

                        _lastSuccess = DateTime.Now;
                    }
                    catch (Exception e)
                    {
                        Logger.Error("Choked on sending the event: {Message}", e.Message);
                        if (e.InnerException != null)
                        {
                            Logger.Error("Inner Exception: {Message}", e.InnerException.Message);
                        }

                        DisplayTimeToExecuteRequest(start, true);
                        var reconnectAttemptMinutes = MonitorProfile.Current.ReConnectAttemptMinutes;
                        if (reconnectAttemptMinutes > 0)
                        {
                            var diff = DateTime.Now - _lastSuccess;
                            if (diff.TotalMinutes >= reconnectAttemptMinutes)
                            {
                                Logger.Error("*** Messages are not getting out! Tring to restarting the service!");
                                Thread.Sleep(2000); // if logger uses a differnt thread(???) then give it time to finish
                                lock (Monitor.FileWriteLock)
                                {
                                    // i know monitor.json is not being written out from any threads in this program
                                    Process.GetCurrentProcess().Kill();
                                }
                            }
                        }
                    }
                }

            } // while
            Logger.Information("Event reporting thread has come to an end");
        }

        private void DisplayTimeToExecuteRequest(DateTime start, bool overrid)
        {
            var end = DateTime.Now;
            var duration = (end - start).TotalMilliseconds;
            duration = duration/1000; // convert to seconds

            // if the duration took longer than 30 seconds then I want to know about it.
            if (duration > 30 || overrid)
            {
                Logger.Information("time in seconds to execute the request: " + duration);
            }
        }

        private Thread? _thread;
        internal void Start()
        {
            _thread = new Thread(Run);
            _thread.Start();
        }
        public void Stop()
        {
            _stopped = true;
            lock (_padlock)
            {
                System.Threading.Monitor.Pulse(_padlock);
            }
            _thread?.Join(10000); // wait a max of 10 seconds to shut this down
        }
    }
}

File: MyMonitorHub.JavaAgent/src/main/java/mymonitorhub/agent/core/EventReportingThread.java
package mymonitorhub.agent.core;

import mymonitorhub.agent.common.MonitorProfile;
import mymonitorhub.agent.common.WebApiCall;
import mymonitorhub.agent.model.Event;
import mymonitorhub.agent.model.EventPacket;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Date;
import java.util.List;

/**
 * Wakes every 60 seconds, drains the event queue, and POSTs to the hub.
 * If events fail to send for longer than reconnect-attempt-minutes the
 * process is killed so the OS can restart it.
 * Mirrors C# EventReportingThread.
 */
class EventReportingThread implements IThreadShutdown {

    private static final Logger log = LoggerFactory.getLogger(EventReportingThread.class);

    private volatile boolean stopped = true;
    private final Object padlock = new Object();
    private Date lastSuccess;
    private Thread thread;

    void start() {
        thread = new Thread(new Runnable() {
            @Override
            public void run() {
                doRun();
            }
        });
        thread.setName("event-reporter");
        thread.setDaemon(true);
        thread.start();
    }

    @Override
    public void stop() {
        stopped = true;
        synchronized (padlock) {
            padlock.notifyAll();
        }
        if (thread != null) {
            try { thread.join(10000); } catch (InterruptedException ignored) {}
        }
    }

    private void doRun() {
        lastSuccess = new Date();
        log.info("Start of Event reporting thread");
        stopped = false;

        while (!stopped) {
            try {
                synchronized (padlock) {
                    padlock.wait(60000);
                }
                if (stopped) break;
            } catch (InterruptedException e) {
                stopped = true;
                break;
            }

            Date start = new Date();
            List<Event> eventList = Monitor.getEventList();

            if (!eventList.isEmpty()) {
                EventPacket packet = new EventPacket(eventList.toArray(new Event[0]));
                try {
                    MonitorProfile profile = MonitorProfile.getCurrent();
                    WebApiCall service = new WebApiCall(profile.getBaseAddress(), MonitorProfile.getTokenManager());
                    service.send(profile.getAccountId(), profile.getDeviceId(), packet);
                    lastSuccess = new Date();
                } catch (Exception e) {
                    log.error("Choked on sending the event: {}", e.getMessage());
                    if (e.getCause() != null) {
                        log.error("Inner Exception: {}", e.getCause().getMessage());
                    }
                    displayDuration(start, true);

                    int reconnectMinutes = MonitorProfile.getCurrent().getReconnectAttemptMinutes();
                    if (reconnectMinutes > 0) {
                        long diffMs = new Date().getTime() - lastSuccess.getTime();
                        long diffMinutes = diffMs / 60000;
                        if (diffMinutes >= reconnectMinutes) {
                            log.error("*** Messages are not getting out! Trying to restart the process!");
                            try { Thread.sleep(2000); } catch (InterruptedException ignored) {}
                            synchronized (Monitor.FILE_WRITE_LOCK) {
                                Runtime.getRuntime().halt(1);
                            }
                        }
                    }
                }
            }
        }
        log.info("Event reporting thread has come to an end");
    }

    private void displayDuration(Date start, boolean force) {
        double seconds = (new Date().getTime() - start.getTime()) / 1000.0;
        if (seconds > 30 || force) {
            log.info("time in seconds to execute the request: {}", seconds);
        }
    }
}
